apps/microtvm/zephyr/demo_runtime/src/main.c

```diff
@@ -22,108 +22,145 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+
+/*
+ * This is a sample Zephyr-based application that contains the logic
+ * needed to upload and control a uTVM-based model via the UART.
+ * This is only intended to be a demonstration, since typically you
+ * will want to incorporate this logic into your own application.
+ */
+
+
 #include <drivers/gpio.h>
 #include <drivers/uart.h>
+#include <fatal.h>
 #include <kernel.h>
 #include <power/reboot.h>
-#include <random/rand32.h>
 #include <stdio.h>
 #include <sys/printk.h>
 #include <sys/ring_buffer.h>
 #include <tvm/runtime/crt/logging.h>
 #include <tvm/runtime/crt/utvm_rpc_server.h>
 #include <unistd.h>
 #include <zephyr.h>
+#include <random/rand32.h>
 
 #ifdef CONFIG_ARCH_POSIX
 #include "posix_board_if.h"
 #endif
 
 #include "crt_config.h"
 
-K_SEM_DEFINE(tx_sem, 0, 1);
-
 static const struct device* tvm_uart;
 
-int write_hook(int c) {
-  uart_poll_out(tvm_uart, c);
-  return 0;
-}
+#ifdef CONFIG_LED
+#define LED0_NODE DT_ALIAS(led0)
+#define LED0 DT_GPIO_LABEL(LED0_NODE, gpios)
+#define LED0_PIN DT_GPIO_PIN(LED0_NODE, gpios)
+#define LED0_FLAGS DT_GPIO_FLAGS(LED0_NODE, gpios)
+static const struct device* led0_pin;
+#endif  // CONFIG_LED
+
+static size_t g_num_bytes_requested = 0;
+static size_t g_num_bytes_written = 0;
 
+// Used by TVM to write serial data to the UART.
 ssize_t write_serial(void* unused_context, const uint8_t* data, size_t size) {
+#ifdef CONFIG_LED
+  gpio_pin_set(led0_pin, LED0_PIN, 1);
+#endif
+  g_num_bytes_requested += size;
+
   for (size_t i = 0; i < size; i++) {
     uart_poll_out(tvm_uart, data[i]);
+    g_num_bytes_written++;
   }
 
+#ifdef CONFIG_LED
+  gpio_pin_set(led0_pin, LED0_PIN, 0);
+#endif
+
   return size;
 }
 
-size_t TVMPlatformFormatMessage(char* out_buf, size_t out_buf_size_bytes, const char* fmt,
-                                va_list args) {
+// This is an error handler which will be invoked if the device crashes.
+// Here, we turn on the LED and spin.
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf) {
+#ifdef CONFIG_LED
+  gpio_pin_set(led0_pin, LED0_PIN, 1);
+#endif
+  for (;;) ;
+}
+
+// Used by TVM when a message needs to be formatted.
+size_t TVMPlatformFormatMessage(char* out_buf, size_t out_buf_size_bytes,
+                                const char* fmt, va_list args) {
   return vsnprintk(out_buf, out_buf_size_bytes, fmt, args);
 }
 
+// Used by TVM when an abort operation occurs.
```

Comment:
nit: "Called by TVM when an internal invariant is violated, and execution cannot continue."